src/hotspot/share/cds/archiveUtils.cpp

```diff
@@ -369,3 +369,22 @@ void ArchiveUtils::log_to_classlist(BootstrapInfo* bootstrap_specifier, TRAPS) {
     }
   }
 }
+
+size_t HeapRootSegments::size_in_bytes(size_t seg_idx) {
+  return objArrayOopDesc::object_size(size_in_elems(seg_idx)) * HeapWordSize;
+}
+
+int HeapRootSegments::size_in_elems(size_t seg_idx) {
+  assert(seg_idx < _count, "In range");
+  if (seg_idx != _count - 1) {
+    return _max_size_in_elems;
+  } else {
+    // Last slice, leftover
+    return _roots_count % _max_size_in_elems;
+  }
+}
+
+size_t HeapRootSegments::segment_offset(size_t seg_idx) {
```

Comment:
Suggestion: add assert on `seg_idx` (as in line 378).

Reply:
Done. Also done in `size_in_bytes`, even though it might be excessive, given the immediate call to `size_in_elems`.